DRA: kubelet: check for duplicate claim names

bart0sh opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Reject NodePrepareResources if the cached claim UID doesn't match resource claim.

#### Which issue(s) this PR fixes:

Fixes #131513

#### Special notes for your reviewer:

This check was proposed in the issue: https://github.com/kubernetes/kubernetes/issues/131513#issuecomment-2842654421

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

/cc @pohly @klueska

pohly commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -234,6 +234,9 @@ func (m *ManagerImpl) prepareResources(ctx context.Context, pod *v1.Pod) error {
 			} else {
+				if claimInfo.ClaimUID != resourceClaim.UID {
+					return fmt.Errorf("claim %s: cached claim UID %s doesn't match resource claim UID %s", klog.KObj(resourceClaim), claimInfo.ClaimUID, resourceClaim.UID)
+				}
```
I prefer a switch statement as in https://github.com/kubernetes/kubernetes/pull/131613#discussion_r2074829636 (less nesting).

Please include the E2E test from that comment, speed it up by not blocking the termination of the second pod, and add unit tests.

bart0sh replied:
> I prefer a switch statement as in https://github.com/kubernetes/kubernetes/pull/131613#discussion_r2074829636 (less nesting).

```
+			default:
 				logger.V(6).Info("Found existing claim info cache entry", "pod", klog.KObj(pod), "podClaim", podClaim.Name, "claim", klog.KObj(resourceClaim), "claimInfoEntry", claimInfo)
```
Wouldn't this change end up logging a new claim as an existing one?

bart0sh replied:
I'd probably swap if/else entries for more readability, wdyt?

```
			// Get a reference to the claim info for this claim from the cache.
			claimInfo, exists := m.cache.get(resourceClaim.Name, resourceClaim.Namespace)
                         if exists {
                                if claimInfo.ClaimUID != resourceClaim.UID {
                                        return fmt.Errorf("claim %s: cached claim UID %s doesn't match resource claim UID %s", klog.KObj(resourceClaim), claimInfo.ClaimUID, resourceClaim.UID)
			        }
                          } else {
                                // Claim info is not cached, add it to the cache.
                                ci, err := newClaimInfoFromClaim(resourceClaim)
			        if err != nil {
				        return fmt.Errorf("claim %s: %w", klog.KObj(resourceClaim), err)
				}
				claimInfo = m.cache.add(ci)
				logger.V(6).Info("Created new claim info cache entry", "pod", klog.KObj(pod), "podClaim", podClaim.Name, "claim", klog.KObj(resourceClaim), "claimInfoEntry", claimInfo)
                          }

```

pohly replied:
> Wouldn't this change end up not logging an existing cache entry when claim UIDs don't match?

Why? Switch cases in Go don't fall through. Besides, the case for "UIDs don't match" returns.

bart0sh replied:
> Why? Switch cases in Go don't fall through. Besides, the case for "UIDs don't match" returns.

My bad. I updated my comment. I'd prefer to see "Found existing claim info cache entry" when claimInfo is found. Your code will show it only if claim UIDs match. And it looks more readable to me with the `if`s. 
I'll update as you've suggested if you insist. It's minor.

> Please include the E2E test from that comment, speed it up by not blocking the termination of the second pod, and add unit tests.

Done, but e2e test succeeds on master in my environment. I thought it shouldn't. Will investigate.

pohly replied:
I find "Found existing claim info cache entry" very confusing if it is not the right one. But I'll defer to a second opinion from @klueska on this and the "is a switch more readable" question.

For comparison, here is how I had written this:
```Go
			switch {
			case !exists:
 				ci, err := newClaimInfoFromClaim(resourceClaim)
 				if err != nil {
 					return fmt.Errorf("claim %s: %w", klog.KObj(resourceClaim), err)
 				}
 				claimInfo = m.cache.add(ci)
 				logger.V(6).Info("Created new claim info cache entry", "pod", klog.KObj(pod), "podClaim", podClaim.Name, "claim", klog.KObj(resourceClaim), "claimInfoEntry", claimInfo)
			case claimInfo.ClaimUID != resourceClaim.UID:
				return fmt.Errorf("old claim with same name %s and different UID %s still exists (previous pod force-deleted?!)", klog.KObj(resourceClaim), resourceClaim.UID)
			default:
 				logger.V(6).Info("Found existing claim info cache entry", "pod", klog.KObj(pod), "podClaim", podClaim.Name, "claim", klog.KObj(resourceClaim), "claimInfoEntry", claimInfo)
 			}
```

I wanted to test on master but then got interrupted and forgot. Yes, it was meant to fail on master. If it doesn't, then further work is needed to find a reproducer for the problem.

bart0sh replied:
I take it back. It was most probably because I didn't restart local cluster. I can't reproduce it anymore: the test fails on master and succeeds on this PR branch as expected.

PTAL again, thanks!

pohly replied:
Looks like delaying shutdown via the `trap ... TERM` doesn't work as intended.

It worked when I increased the `sleep 30` to `sleep 60`:

```
  [FAILED] Only new claim should be prepared now because new pod is running.
  Expected
      <[]app.ClaimID | len:1, cap:1>: [
          {
              Name: "external-claim",
              UID: "4fb4f94b-1731-40c7-aba0-7df70fe4cbd3",
          },
      ]
  to equal
      <[]app.ClaimID | len:1, cap:1>: [
          {
              Name: "external-claim",
              UID: "55373d34-2ff4-446c-a990-78ef6bf860f0",
          },
      ]
  In [It] at: /nvme/gopath/src/k8s.io/kubernetes/test/e2e/dra/dra.go:223 @ 05/06/25 20:27:51.352
```

pohly replied:
This variant of the test fails reliably for me on master:
```patch
diff --git a/test/e2e/dra/dra.go b/test/e2e/dra/dra.go
index e75f214c3de..8a013e12960 100644
--- a/test/e2e/dra/dra.go
+++ b/test/e2e/dra/dra.go
@@ -47,6 +47,7 @@ import (
 	"k8s.io/dynamic-resource-allocation/resourceslice"
 	"k8s.io/klog/v2"
 	"k8s.io/kubernetes/pkg/features"
+	testdriverapp "k8s.io/kubernetes/test/e2e/dra/test-driver/app"
 	"k8s.io/kubernetes/test/e2e/feature"
 	"k8s.io/kubernetes/test/e2e/framework"
 	e2edaemonset "k8s.io/kubernetes/test/e2e/framework/daemonset"
@@ -169,6 +170,61 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
 			}
 		})
 
+		ginkgo.It("blocks new pod after force-delete", func(ctx context.Context) {
+			// The problem with a force-deleted pod is that kubelet
+			// is not necessarily done yet with tearning down the
+			// pod at the time when the pod and its claim are
+			// already removed. The user can replace the claim and
+			// pod with new instances under the same name.  The
+			// kubelet then needs to detect that the new claim is
+			// not the same as the one that kubelet currently works
+			// on and that the new pod cannot start until the old
+			// one is torn down.
+			//
+			// This test delays termination of the first pod to ensure
+			// that the race goes bad (old pod pending when new one
+			// arrives) and always schedules to the same node.
+			claim := b.externalClaim()
+			pod := b.podExternal()
+			pod.Spec.TerminationGracePeriodSeconds = ptr.To(int64(60))
+			oldCommand := pod.Spec.Containers[0].Command
+			pod.Spec.Containers[0].Command = []string{"/bin/sh", "-c", "trap '/bin/sleep 60; exit 10' TERM; while true; do /bin/sleep 1; done"}
+			node := nodes.NodeNames[0]
+			pod.Spec.NodeSelector = map[string]string{"kubernetes.io/hostname": node}
+			oldClaim := b.create(ctx, claim, pod)[0].(*resourceapi.ResourceClaim)
+			b.testPod(ctx, f, pod)
+
+			ginkgo.By("Force-delete claim and pod")
+			forceDelete := metav1.DeleteOptions{GracePeriodSeconds: ptr.To(int64(0))}
+			framework.ExpectNoError(b.f.ClientSet.CoreV1().Pods(b.f.Namespace.Name).Delete(ctx, pod.Name, forceDelete))
+
+			// The pod should get deleted immediately.
+			_, err := b.f.ClientSet.CoreV1().Pods(b.f.Namespace.Name).Get(ctx, pod.Name, metav1.GetOptions{})
+			if !apierrors.IsNotFound(err) {
+				framework.Failf("Expected 'not found' error, got: %v", err)
+			}
+
+			// The claim may take a bit longer because of the allocation and finalizer.
+			framework.ExpectNoError(b.f.ClientSet.ResourceV1beta1().ResourceClaims(b.f.Namespace.Name).Delete(ctx, claim.Name, forceDelete))
+			gomega.Eventually(ctx, func(ctx context.Context) (*resourceapi.ResourceClaim, error) {
+				claim, err := b.f.ClientSet.ResourceV1beta1().ResourceClaims(b.f.Namespace.Name).Get(ctx, claim.Name, metav1.GetOptions{})
+				if apierrors.IsNotFound(err) {
+					return nil, nil
+				}
+				return claim, err
+			}).Should(gomega.BeNil())
+			gomega.Expect(b.driver.Nodes[node].GetPreparedResources()).Should(gomega.Equal([]testdriverapp.ClaimID{{Name: oldClaim.Name, UID: oldClaim.UID}}), "Old claim should still be prepared.")
+
+			ginkgo.By("Re-creating the same claim and pod")
+			pod.Spec.Containers[0].Command = oldCommand
+			newClaim := b.create(ctx, claim, pod)[0].(*resourceapi.ResourceClaim)
+			b.testPod(ctx, f, pod)
+
+			// The pod must not have started before NodeUnprepareResources was called for the old one,
+			// i.e. what is prepared now must be the new claim.
+			gomega.Expect(b.driver.Nodes[node].GetPreparedResources()).Should(gomega.Equal([]testdriverapp.ClaimID{{Name: newClaim.Name, UID: newClaim.UID}}), "Only new claim should be prepared now because new pod is running.")
+		})
+
 		ginkgo.It("must call NodePrepareResources even if not used by any container", func(ctx context.Context) {
 			pod, template := b.podInline()
 			for i := range pod.Spec.Containers {
```

It fails after 14 seconds.

pohly replied:
This is still a bit odd, because 30 seconds should have been long enough.

pohly replied:
I suspect the kubelet ignores the `terminationGracePeriodSeconds: 60` after the client's force-delete of the pod and tries to shut down the old pod immediately. Then it becomes a matter of timing whether the new pod arrives quickly enough. We might need a loop to trigger the bad outcome of the race more reliably.

bart0sh replied:
I still see UnprepareResources called for old claim and pod. I thought it shouldn't happen if grace period is 0. I probably misunderstood the case described in the issue.

pohly replied:
It's a race. Sometimes the kubelet manages to shut down the old pod before it sees the new one, sometimes it doesn't.

pohly replied:
Grace period 0 doesn't mean that the kubelet doesn't shut down the pod and doesn't call NodeUnprepareResources. It still sees the pod deletion and tries to act accordingly. But in the meantime the new pod already gets scheduled.

bart0sh replied:
> This variant of the test fails reliably for me on master:

It failed at least once in the PR branch in 11.8 seconds:
```
  [FAILED] Old claim should still be prepared.
  Expected
      <[]app.ClaimID | len:0, cap:0>: nil
  to equal
      <[]app.ClaimID | len:1, cap:1>: [
          {
              Name: "external-claim",
              UID: "1ab33383-8e07-4c7c-ab2d-20a700d54565",
          },
      ]
  In [It] at: k8s.io/kubernetes/test/e2e/dra/dra.go:328 @ 05/07/25 10:43:27.23
```

pohly replied:
Same problem: the assumption that the kubelet keeps the pod running for the duration of `terminationGracePeriod` is wrong.

In that failure, the kubelet cleaned up before the test reached that check.

bart0sh replied:
For orphaned pods (pods that are not reported by API server, but reported by runtime) grace period is overriden to 1 sec: https://github.com/kubernetes/kubernetes/blob/release-1.33/pkg/kubelet/kubelet_pods.go#L1338
I don't see an easy way to change this.

So, what should we do with this PR?

pohly replied:
The fix in manager.go is needed.

Testing the race in the E2E test is a bit harder than hoped. Add a for loop around creating/deleting with something like 10 to 20 iterations, then check that the number of NodePrepareResources calls matches the number of iterations. At least for me that failed fairly reliably when done in the shell test.

bart0sh replied:
That could be a time consuming and probably flaky test. Are you sure we need it?

pohly replied:
The test is potentially unreliable in the sense that there might be a false positive. That's not a problem. The test correctly identified a problem in your branch, which IMHO makes it worthwhile.

How long does it run for 20 iterations? Let's check before we decide anything.

bart0sh replied:
@pohly I created an artificial delay by failing `NodeUnprepareResources` and now the test is more stable. 
Would this be ok to use this approach from your point of view. Here is the change:
```
diff --git a/test/e2e/dra/dra.go b/test/e2e/dra/dra.go
index 1563a57e4e7..9cd581f54b2 100644
--- a/test/e2e/dra/dra.go
+++ b/test/e2e/dra/dra.go
@@ -298,10 +298,6 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
                        // arrives) and always schedules to the same node.
                        claim := b.externalClaim()
                        pod := b.podExternal()
-                       savedGracePeriod := pod.Spec.TerminationGracePeriodSeconds
-                       savedCommand := pod.Spec.Containers[0].Command
-                       pod.Spec.TerminationGracePeriodSeconds = ptr.To(int64(60))
-                       pod.Spec.Containers[0].Command = []string{"/bin/sh", "-c", "trap 'sleep 30; exit 10' TERM; while true; do sleep 1; done"}
                        node := nodes.NodeNames[0]
                        pod.Spec.NodeSelector = map[string]string{"kubernetes.io/hostname": node}
                        oldClaim := b.create(ctx, claim, pod)[0].(*resourceapi.ResourceClaim)
@@ -311,6 +307,10 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
                        forceDelete := metav1.DeleteOptions{GracePeriodSeconds: ptr.To(int64(0))}
                        framework.ExpectNoError(b.f.ClientSet.CoreV1().Pods(b.f.Namespace.Name).Delete(ctx, pod.Name, forceDelete))
 
+                       // Fail NodeUnprepareResources to simulate long grace period
+                       unprepareResources := MethodInstance{node, NodeUnprepareResourcesMethod}
+                       driver.Fail(unprepareResources, true)
+
                        // The pod should get deleted immediately.
                        _, err := b.f.ClientSet.CoreV1().Pods(b.f.Namespace.Name).Get(ctx, pod.Name, metav1.GetOptions{})
                        if !apierrors.IsNotFound(err) {
@@ -328,10 +328,9 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
                        }).Should(gomega.BeNil())
                        gomega.Expect(b.driver.Nodes[node].GetPreparedResources()).Should(gomega.Equal([]testdriverapp.ClaimID{{Name: oldClaim.Name, UID: oldClaim.UID}}), "Old claim should still be prepared.")
 
+                       driver.Fail(unprepareResources, false)
+
                        ginkgo.By("Re-creating the same claim and pod")
-                       // Restore the original values to speed up the test.
-                       pod.Spec.TerminationGracePeriodSeconds = savedGracePeriod
-                       pod.Spec.Containers[0].Command = savedCommand
                        newClaim := b.create(ctx, claim, pod)[0].(*resourceapi.ResourceClaim)
                        b.testPod(ctx, f, pod)
```

The test is much faster now:
```
Ran 1 of 6998 Specs in 17.043 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 6997 Skipped
--- PASS: TestE2E (17.22s)
PASS

real	0m18.016s
user	0m4.275s
sys	0m0.473s
```
Previously it was around 1m30s.

bart0sh replied:
> How long does it run for 20 iterations? Let's check before we decide anything.

One iteration takes 1.5 min in my setup. CI infra is usually much slower.

pohly replied:
> I created an artificial delay by failing NodeUnprepareResources and now the test is more stable.

That sounds good.

pohly replied:
xref: https://github.com/kubernetes/kubernetes/pull/131617#discussion_r2084515571

Can we at least emit different log message for the happy path (code found the right claim info) and the unhappy one (found it, but it's the wrong one)?

Logging before returning an error is usually an anti-pattern, but I seem to be in the minority here.

bart0sh replied:
> Logging before returning an error is usually an anti-pattern, but I seem to be in the minority here.

I agree with you here. Fixed. PTAL.

bart0sh replied:
@klueska Does it look OK to you now?

bart0sh commented:
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/retest

bart0sh committed:
DRA: genTestClaimInfo: add claimUID parameter

bart0sh committed:
DRA: test mismatched claim UIDs

bart0sh commented:
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented:
/test pull-kubernetes-unit

bart0sh commented on test/e2e/dra/dra.go:
```diff
@@ -281,6 +282,71 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+			// Keep blocking NodeUnprepareResources for the first pod
+			// until the second one is pending on the node.
+			// This should be enough to trigger the race.
+			go func(podName, namespace string) {
```
It fails from time to time with the diff that I've shown to you earlier. I tried to unblock just before `b.testPod` call, but it was still unstable.

bart0sh replied:
Just tried one more time with this diff:
```
diff --git a/test/e2e/dra/dra.go b/test/e2e/dra/dra.go
index 73dc1e5ab78..de50e049494 100644
--- a/test/e2e/dra/dra.go
+++ b/test/e2e/dra/dra.go
@@ -334,12 +334,12 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
                        // Keep blocking NodeUnprepareResources for the first pod
                        // until the second one is pending on the node.
                        // This should be enough to trigger the race.
-                       go func(podName, namespace string) {
-                               e2epod.WaitForPodCondition(ctx, f.ClientSet, namespace, podName, "pending", podStartTimeout, func(pod *v1.Pod) (bool, error) {
-                                       return pod.Status.Phase == v1.PodPending, nil
-                               })
-                               driver.Fail(unprepareResources, false)
-                       }(pod.Name, pod.Namespace)
+                       //go func(podName, namespace string) {
+                       e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name, "pending", podStartTimeout, func(pod *v1.Pod) (bool, error) {
+                               return pod.Status.Phase == v1.PodPending, nil
+                       })
+                       driver.Fail(unprepareResources, false)
+                       //}(pod.Name, pod.Namespace)
 
                        b.testPod(ctx, f, pod)
```

The test sometimes takes around 15 sec with no error in the kubelet log, which means that Kubelet was able to call UnprepareResources for the first pod before the first call of PrepareResources for the second pod.

pohly replied:
I don't think this has anything to do with the goroutine.

Is `pod.Status.Phase == v1.PodPending` set by the scheduler or the kubelet? I think it's the scheduler. It's documented as "This includes time before being bound to a node... ", so this can be reached before even scheduling the pod.

In that case the test lifts the block for the first pod before it is certain that the kubelet has seen the second pod and therefore we still have a race.

What the test needs to wait for is a signal that the kubelet has tried to start the second pod. We could wait for the event with the desired error message. If we wait long enough, eventually we should see it, even if some events get dropped (the usual reason why tests should not rely on reliable event delivery). Or does the kubelet write something into the pod status after encountering the incorrect claim?

The test then doesn't work on master, but that's okay.

bart0sh replied:
Implemented without running go-routine. PTAL.

bart0sh commented:
/retest

pohly commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -226,15 +226,18 @@ func (m *ManagerImpl) prepareResources(ctx context.Context, pod *v1.Pod) error {
+				if claimInfo.ClaimUID != resourceClaim.UID {
+					return fmt.Errorf("claim %s: cached claim UID %s doesn't match resource claim UID %s", klog.KObj(resourceClaim), claimInfo.ClaimUID, resourceClaim.UID)
+				}
+			} else {
```
I still think a switch statement would make the three cases more obvious, see https://github.com/kubernetes/kubernetes/pull/131617#discussion_r2075902756

Also let's add the `(pod force-deleted ?!)` hint to the error. If a user sees this as it is now, they won't know why it occurred. Force-deletion is the most likely explanation.

cc @klueska

bart0sh replied:
Should the error message contain cached UID? The way it looks in your proposed change is confusing:
```
return fmt.Errorf("old claim with same name %s and different UID %s still exists (previous pod force-deleted?!)", klog.KObj(resourceClaim), resourceClaim.UID)
```

bart0sh replied:
I'd propose this variant:
```
fmt.Errorf("old claim with same name %s and different UID %s still exists (previous pod force-deleted?!)", claimInfo.ClaimName, claimInfo.ClaimUID)
```
It still sounds a little bit confusing to me, but at least it includes old claim name and uid.

pohly commented on test/e2e/dra/dra.go:
```diff
@@ -281,6 +285,81 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
 			testContainerEnv(ctx, f, pod, pod.Spec.Containers[0].Name, true, container0Env...)
 			testContainerEnv(ctx, f, pod, pod.Spec.Containers[1].Name, true, container1Env...)
 		})
+		ginkgo.It("blocks new pod after force-delete", func(ctx context.Context) {
```
Please add an empty line above it for readability.

pohly commented on test/e2e/dra/dra.go:
```diff
@@ -281,6 +285,81 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
 		})
+		ginkgo.It("blocks new pod after force-delete", func(ctx context.Context) {
+			// The problem with a force-deleted pod is that kubelet
+			// is not necessarily done yet with tearning down the
```
```suggestion
			// is not necessarily done yet with tearing down the
```

pohly commented on test/e2e/dra/dra.go:
```diff
@@ -281,6 +285,81 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+			// one is torn down.
+			//
+			// This test delays termination of the first pod to ensure
+			// that the race goes bad (old pod pending when new one
```
```suggestion
			// that the race goes bad (old pod pending shutdown when new one
```

pohly commented on test/e2e/dra/dra.go:
```diff
@@ -281,6 +285,81 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+				"reason":                   events.FailedPrepareDynamicResources,
+			}.AsSelector().String()
+			gomega.Eventually(ctx, func(ctx context.Context) error {
+				return e2eevents.WaitTimeoutForEvent(
```
Why are you wrapping a function which polls itself in `gomega.Eventually`? Wouldn't it be simpler to call `e2eevents.WaitTimeoutForEvent` once with the overall timeout that you want?

Speaking of that timeout, what we want here is something that is long enough that the kubelet should have emitted the event multiple times (in case that we missed an event). `3 * f.Timeouts.PodDelete`? A comment about the choice of the timeout for posterity would be good.

pohly committed:
e2e: DRA: test 2 claims with the same name

Co-Authored-By: Ed Bartosh <eduard.bartosh@intel.com>

pohly commented on test/e2e/dra/dra.go:
```diff
@@ -281,6 +285,85 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+				pod.Namespace,
+				expectedEvent,
+				fmt.Sprintf("old claim with same name %s and different UID %s still exists", klog.KObj(oldClaim), oldClaim.UID),
+				framework.PodStartTimeout*2))
```
`WaitTimeoutForEvent` is one of those wait functions which fails poorly in case of a timeout (here when run against master):
```
  [FAILED] got error while getting events: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline
  In [It] at: /nvme/gopath/src/k8s.io/kubernetes/test/e2e/dra/dra.go:351 @ 05/08/25 13:37:07.607
```

But improving it falls under https://github.com/kubernetes/kubernetes/issues/106575, not this PR.

When replacing it with a 10 second sleep, we get the expected failure on master:
```
  [FAILED] Only new claim should be prepared now because new pod is running.
  Expected
      <[]app.ClaimID | len:1, cap:1>: [
          {
              Name: "external-claim",
              UID: "01f09231-f6e0-41d4-aeec-8927ddd189b2",
          },
      ]
  to equal
      <[]app.ClaimID | len:1, cap:1>: [
          {
              Name: "external-claim",
              UID: "a5fa6c6f-9800-470b-8c24-1b79614836b0",
          },
      ]
  In [It] at: /nvme/gopath/src/k8s.io/kubernetes/test/e2e/dra/dra.go:363 @ 05/08/25 14:07:29.737
```

With the fix from this, branch, it passes in ~90 seconds. Using `WaitTimeoutForEvent` doesn't make much difference, most of the delays seem to come from the kubelet retry loop. But that's okay.

/lgtm
/assign @klueska 

For approval.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -226,14 +226,17 @@ func (m *ManagerImpl) prepareResources(ctx context.Context, pod *v1.Pod) error {
 			claimInfo, exists := m.cache.get(resourceClaim.Name, resourceClaim.Namespace)
-			if !exists {
+			switch {
+			case !exists:
 				ci, err := newClaimInfoFromClaim(resourceClaim)
 				if err != nil {
 					return fmt.Errorf("claim %s: %w", klog.KObj(resourceClaim), err)
 				}
 				claimInfo = m.cache.add(ci)
 				logger.V(6).Info("Created new claim info cache entry", "pod", klog.KObj(pod), "podClaim", podClaim.Name, "claim", klog.KObj(resourceClaim), "claimInfoEntry", claimInfo)
-			} else {
+			case claimInfo.ClaimUID != resourceClaim.UID:
+				return fmt.Errorf("old claim with same name %s and different UID %s still exists (previous pod force-deleted?!)", klog.KObj(resourceClaim), claimInfo.ClaimUID)
+			default:
 				logger.V(6).Info("Found existing claim info cache entry", "pod", klog.KObj(pod), "podClaim", podClaim.Name, "claim", klog.KObj(resourceClaim), "claimInfoEntry", claimInfo)
 			}
```
I think we should capture the last log even if claimInfo.ClaimUID != resourceClaim.UID.

As such, I think you can leave the existing code as is, and just add the following:
```
if claimInfo.ClaimUID != resourceClaim.UID {
	return fmt.Errorf("old claim with same name %s and different UID %s still exists (previous pod force-deleted?!)"
}
```
In the case where the claim didn't previously exist in the cache, we will have just inserted it, so we know this check can never be true.

bart0sh replied:
That was my initial implementation that triggered [this discussion](https://github.com/kubernetes/kubernetes/pull/131617#discussion_r2074876967).

Let me know if you still want me to do that.

@pohly

klueska replied:
I actually don't feel too strongly either way. If @pohly feels strongly about a switch statement then that is fine with me (I had missed his comments since they weren't in the diff anymore when I looked after you had updated).

bart0sh replied:
So, what to do? I don't feel strongly about this, just want it to be merged.

pohly replied:
I defer to @klueska. He needs to approve anyway.

bart0sh committed:
DRA: kubelet: check for duplicate claim names

Reject NodePrepareResources if the cached claim UID doesn't
match resource claim.

bart0sh commented:
/triage accepted
/priority important-longterm

bart0sh commented:
/retest

klueska commented:
/approve
/lgtm

bart0sh commented:
/retest
